tailcfg/tailcfg.go

```diff
@@ -1369,3 +1373,82 @@ type SetDNSRequest struct {
 	// Value is the value to add.
 	Value string
 }
+
+// SSHPolicy is the policy for how to handle incoming SSH connections
+// over Tailscale.
+type SSHPolicy struct {
+	// Rules are the rules to process for an incoming SSH
+	// connection. The first matching rule takes its action and
+	// stops processing further rules.
+	Rules []*SSHRule
+}
+
+// An SSH rule is a match predicate and associated action for an incoming SSH connection.
+type SSHRule struct {
+	// RuleExpires, if non-nil, is when this rule expires.
+	//
+	// For example, a (principal,sshuser) tuple might be granted
+	// prompt-free SSH access for N minutes, so this rule would be
+	// before a expiration-free rule for the same principal that
+	// required an auth prompt.  This permits the control plane to
+	// be out of the path for already-authorized SSH pairs.
+	//
+	// Once a rule matches, the lifetime of any accepting connection
+	// is subject to the SSHAction.SessionExpires time, if any.
+	RuleExpires *time.Time `json:"ruleExpires,omitempty"`
+
+	// Principal matches an incoming connection. If the connection
+	// matches anything in this list and also matches SSHUser,
+	// then Action is applied.
+	Principal []*SSHPrincipal `json:"principal"`
+
+	// SSHUser are the SSH users that this rule matches.  It is a
+	// map from either ssh-user|"*" => local-user.  The map must
+	// contain a key for either ssh-user or, as a fallback, "*" to
+	// match anything. If it does, the map entry's value is the
+	// actual user that's logged in.
+	SSHUser map[string]string `json:"sshUser"`
+
+	// Action is the outcome to task.
+	// A nil or invalid action means to deny.
+	Action *SSHAction `json:"action"`
+}
+
+// SSHPrincipal is either a particular node or a user on any node.
+// At most one field should be non-zero specified.
+type SSHPrincipal struct {
+	Node StableNodeID `json:"node,omitempty"`
+	User UserID       `json:"user,omitempty"`
```

Comment:
We have a StableUserID on the server, I think we should add it to tailcfg and use it here. I'd rather not introduce a use for the old UserID here.

Reply:
adjusted